### PR TITLE
[enhancement](nereids) support case when for TPC-H

### PR DESCRIPTION
**Pull request by yinzhijian**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:
support case when for TPC-H
for example:
CASE [expression] WHEN [value] THEN [expression] ... ELSE [expression] END
or
CASE  WHEN [predicate] THEN [expression] ... ELSE [expression] END
## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/WhenClause.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * captures info of a single WHEN expr THEN expr clause.
+ */
+public class WhenClause extends Expression implements BinaryExpression {
+    public WhenClause(Expression operand, Expression result) {
+        super(ExpressionType.WHEN_CLAUSE, operand, result);
+    }
+
+    @Override
+    public String toSql() {
+        return "when " + left().toSql() + " then " + right().toSql();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/WhenClause.java:39]: nit:
```suggestion
        return "WHEN " + left().toSql() + " THEN " + right().toSql();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java [hunk 2]
```diff
@@ -213,6 +218,23 @@ public Expr visitStringRegexPredicate(StringRegexPredicate stringRegexPredicate,
...
+    public Expr visitCaseWhen(CaseWhen caseWhen, PlanTranslatorContext context) {
+        List<CaseWhenClause> caseWhenClauses = new ArrayList<>();
+        for (WhenClause whenClause : caseWhen.getWhenClauses()) {
+            caseWhenClauses.add(new CaseWhenClause(
+                    whenClause.left().accept(this, context),
+                    whenClause.right().accept(this, context)
+            ));
+        }
+        Expr elseExpr = null;
+        Optional<Expression> defaultValue = caseWhen.getDefaultValue();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java:231]: nit: Optional is unnecessary

File: fe/fe-core/src/test/java/org/apache/doris/nereids/trees/expressions/ExpressionParserTest.java [hunk 3]
```diff
@@ -137,4 +137,13 @@ public void testSortClause() throws Exception {
         String sort1 = "select a from test order by 1";
         assertSql(sort1);
     }
+
+    @Test
+    public void testCaseWhen() throws Exception {
+        String caseWhen = "select case a when 1 then 2 else 3 end from test";
+        assertSql(caseWhen);
+
+        String caseWhen2 = "select case when a = 1 then 2 else 3 end from test";
```
Comment (qzsee) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/trees/expressions/ExpressionParserTest.java:146]: I think should another add some ut case. add complex expr after then and else?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CaseWhen.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+ * Each When/Then is stored as two consecutive children (whenExpr, thenExpr).
+ * If a case expr is given, convert it to equalTo(caseExpr, whenExpr) and set it to whenExpr.
+ * If an else expr is given then it is the last child.
+ */
+public class CaseWhen extends Expression {
+    /**
+     * If default value exists, then defaultValueIndex is the index of the last element in children,
+     * otherwise it is -1
+     */
+    private final int defaultValueIndex;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CaseWhen.java:44]: utilize index to get different part of children is ugly, we should refactor to remove children from AbstractTreeNode and provide children by concrete child class later

**Review by 924060929 — APPROVED** [review]
LGTM

File: fe/fe-core/src/test/java/org/apache/doris/nereids/trees/expressions/ExpressionParserTest.java (hunk 3 above)
Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/trees/expressions/ExpressionParserTest.java:146]: @yinzhijian add ut later

**Pull request merged**